### ISSUE TITLE
If two or more files have the same name, show part of their path in the tab

### ISSUE THREAD

**Issue by tgross35** [WerWolv/ImHex#1162]
### What feature would you like to see?

To my knowledge, there isn't any way to differentiate which files you have open if they have the same name.

<img width="159" alt="image" src="https://github.com/WerWolv/ImHex/assets/13724985/48161e08-e1a0-4915-9450-0168a39303dc">

My suggestion is to display part of uncommon path to make it easier to see at a glance - maybe in subtle text. For example, with the two paths:

```
foo/Data
bar/Data
```

Titles could be something like "Data _foo/_" and "Data _bar/_". Or for paths with more common ancestors:

```
foo/bar/Data
baz/bar/Data
qux/Data
```

Something like "Data _foo/..._", "Data _baz/..._" and "Data _qux/_".

A few other related ideas:

- Show the full path when you hover over the tab (or at least the last 40ish characters)
- Show the path in the bottom or top bar, maybe even with an "open containing directory" icon.
- Show part of the path in the "recent files" section of the welcome page, or at least on hover (It seems like maybe it only stores the most recent file with a given name?)

### How will this feature be useful to you and others?

I can't be the only one who seems to always wind up working with a lot of different files with the same name - it's just pretty easy to get confused.

### Request Type

- [ ] I can provide a PoC for this feature or am willing to work on it myself and submit a PR

**Comment by paxcut** [comment]
I agree the tool tip displaying when tab is hovered should display it path, but the name of the tabs should not be changed. 

If it is confusing to work with several files that have the same name then reconsider the reasons they use the same name to begin with. Files have the same name if each one is unique to its folder like two README files.  if you are going to work with several READMEs add a suffix (_id) to distinguish them using the name of the project they belong to. 

Most software (eg notepad++) handles it that way (tool-tip with path but tabs with same name). 
this is just my opinion though.

**Comment by tgross35** [comment]
> If it is confusing to work with several files that have the same name then reconsider the reasons they use the same name to begin with. Files have the same name if each one is unique to its folder like two README files. if you are going to work with several READMEs add a suffix (_id) to distinguish them using the name of the project they belong to.

The archive formats I work with just name it all the same, I don't do it that way by choice. Unique names wouldn't make sense anyway, the structure is something like this:

```
root/
    Data
     item1/
        Data
        Storage
        Meta
     item2/
        Data
        Storage
     item3/
        Data
        Storage
...
```

**Comment by paxcut** [comment]
While you didn't choose to name the files the same you are choosing to open them with an app that creates tabs with the name of the file.  so either open them sequentially in order or rename them as Data_root, Data_item1, Storage_item1, Data_item2, ... but create copies to open in imhex. 

I am not aware of any software that uses tabs with file names that adds parts of the path to the name of the tab to differentiate tabs with the same name. A tool tip is often displayed with a full path which will identify the file uniquely and this is the way all the apps i have tested deal with it.

**Comment by iTrooz** [comment]
> I am not aware of any software that uses tabs with file names that adds parts of the path to the name of the tab to differentiate tabs with the same name

VSCode and Kate (the only two I have tested) do this ^^ and it seems pretty standard behaviour to me

![image](https://github.com/WerWolv/ImHex/assets/42669835/cf584622-c017-481e-a9ef-1ad3aff6e77b)
![image](https://github.com/WerWolv/ImHex/assets/42669835/b5a66ddd-a005-4362-87bf-e77732c4cd44)

I personally think it is a good idea, but we seem to disagree. @WerWolv What do you think ?

**Comment by WerWolv** [comment]
I'm all for adding extra information to the tabs if two tabs have the same name. 
It will probably have to look like Kate though as we don't have any formatting options available inside the tabs. 

Also this will need to be handled for all the different providers. For example a file provider would need to display part of the path, the GDB provider maybe the IP and Port and so on.
Maybe it would make sense to just give people the ability to assign "Nicknames" to tabs that get stored with the project file?

**Comment by paxcut** [comment]
I stand corrected about the existence of software that handles equally named tabs by changing their names. I disagree that what iTrooth found is the 'standard behavior', as this was already clearly noted in my post which included a  big counter-example (notepad++) which he completely ignored in his post. 

Just like him, I tested two programs and both kept the tab names equal. His findings did call for a more systematic search for how this is handled which yielded a mixture of results. Some, like Clion, change both tab names with a partial path before the name, not after the name. Others like beyond compare, do the same thing but add a more complete path before the name and it makes more sense to add path information before the name than after but either way works for me. 

So it appears that a standard behavior for this feature does not exist and the lack of examples in the op made me think that the standard behavior was as described in my post but I made sure to point out that it was only an opinion because claiming that something is standard behavior requires a more exhaustive search than just two apps. I think that everybody would agree that  claiming something seems to be standard behavior while ignoring counter-examples of popular apps given in previous posts will indeed  cause disagreement because the claim has already been proven wrong.

In light of the new information I have changed my mind about what should be done in this case and modifying the name is the way to go. When asking for features it is important to give examples of how other popular apps handle it because it helps to explain the logic behind the request and it also helps to see the feature in a more complete way.

I still believe than a tool-tip that shows the same information that is already displayed, when more complete information is available, is not very useful so hovering the tabs should show the full path and not the file name as it does now.

**Comment by iTrooz** [comment]
I said "it seems", and that's really how I work. I'm going to find few examples, and say "it seems that this is always true", and wait to be proven wrong.
That said, I admit I didn't really consider the notepad++ example, even though I remember reading it. sorry about that.


> I still believe than a tool-tip that shows the same information that is already displayed, when more complete information is available, is not very useful so hovering the tabs should show the full path and not the file name as it does now.

Did you know you could show the full path by pressing "Shift" ? I added a text to introduce that feature to everyone earlier today: https://github.com/WerWolv/ImHex/pull/1171

As to how to modify the name to differentiate them (what WerWolv said), I really don't know what such an algorithm would look like. Maybe we should take a peek to see how other software do it.

**Comment by paxcut** [comment]
@iTrooz I work the exact same way when testing features that have no examples of apps mentioned in op.
I failed to provide enough emphasis in my mentioning of notepad++ and that may have have cause you to overlook it understandingly so I should be the one apologizing.
Also note that I explicitly use the verb 'seems' when I quote your claims because I tend to do the same if I don't explicitly state that only an opinion is being offered.

I wasn't aware of the shift key shortcut and saw your pr a bit too late after i wrote my previous post in this thread. My opinion remains unchanged on this subject.

It doesn't seem necessary (or at least intuitive) to have to add a tip on a tool-tip to explain how to obtain the information being sought after when it can just be provided now. Another way to say this is that currently the tool-tip feature on tab names does absolutely nothing but show whats already being shown on the tab so its very existence is at best questionable. To add insult to injury we ask users to hold the shift key suggesting it was their fault that we wasted the last few seconds of their time for not knowing that ahead of time which seems rather unreasonable as your and my own experiences indicate. Like me (and probably you though idrk), most people would shrug that off and move on so it is fine either way.

Now to the real question to be addressed. How exactly do we modify the tab label when a new item has the same name? I don't have  a strong predilection but logic indicates the change should mimic a full path as close as possible and should be minimal. That's easier said than done though so as a suggestion I delineate a possible approach.
1)  break all paths down to their components.
2) find the first component that is different in each pair 
3) using '..' as notation for 'whatever should go here' which may include one or more common components compose both names with [../]different_component/[../]/name_of_file where the [] indicates that they may or may no be necessary depending on the locations of the different component. This is not a complete solution as there are cases that are not covered like the one i show below.
some examples
```
c:/A/Readme and c:/B/Readme
'''
 turn into../A/Readme and ../B/Readme
For the op example and assuming the rest of the paths are the same and just the drive root label and adding a tricky entry
```
c:/foo/bar/Data
c:/baz/bar/Data
c:/qux/Data
c:/foo/bar/qux/Data
```
we get should get the following but i am not sure how 
../foo/../Data ../bar/../Data  ../qux/Data and ../foo/../qux/Data
There may be other fringe cases not considered and hard to deal with that become apparent as the code is being written. I consider this kind of feature as hard to code which was one of the reasons for my original suggestion which is what notepad++ does. Frankly, the more I think about it  the less sure I become about what the best overall approach should be but I am willing to continue the discussion and continue to believe we are in the right path.
